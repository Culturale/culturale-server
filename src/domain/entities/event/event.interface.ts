import type { IChat } from '~/domain/entities/chat/chat.interface';
import type { MongoId } from '~/types/types';

import type { IUser } from '../user';

export interface IEvent {
  id: MongoId;
  codi: number;
  denominacio: string;
  descripcio: string;
  dataIni: Date;
  dataFi: Date;
  horari?: string;
  adress?: string;
  url: string;
  chat: IChat;
  participants: IUser[];

  updateParticipant: (newParticipant: IUser) => void;
}
