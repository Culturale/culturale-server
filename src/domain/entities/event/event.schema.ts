import { Schema, model } from 'mongoose';

import type { IEvent } from './event.interface';

const CategoriaEnum = {
  ActivitatsVirtuals: 'agenda:categories/activitats-virtuals',
  Exposicions: 'agenda:categories/exposicions',
  Concerts: 'agenda:categories/concerts',
  Teatre: 'agenda:categories/teatre',
  FestivalsIMostres: 'agenda:categories/festivals-i-mostres',
  RutesIVisites: 'agenda:categories/rutes-i-visites',
  Infantil: 'agenda:categories/infantil',
  Festes: 'agenda:categories/festes',
  Conferencies: 'agenda:categories/conferencies',
  FiresIMercats: 'agenda:categories/fires-i-mercats',
  Dansa: 'agenda:categories/dansa',
  Cicles: 'agenda:categories/cicles',
};


const EventSchema = new Schema({
  adress: { required: true, type: String },
  chat: { required: true, type: Schema.Types.ObjectId, ref: 'Chat' },
  codi: { required: true, type: Number },
  dataFi: { required: true, type: Date },
  dataIni: { required: true, type: Date },
  denominacio: { required: true, type: String },
  descripcio: { required: true, type: String },
  horari: { required: true, type: String },
  url: { required: true, type: String },
  valoracions: {
    required: false,
    type: [Schema.Types.ObjectId],
    model: 'Review',
  },
  participants: {
    required: false,
    type: [Schema.Types.ObjectId],
    model: 'User',
  },
  assistents: {
      required: false,
      type: [Schema.Types.ObjectId],
      model: 'User',
    },
  lat: { required: false, type: Number },
  long: { required: false, type: Number },
  price: { required: false, type: String },
  photo: { required: false, type: String },
  categoria: { required: true, type: String, enum: Object.values(CategoriaEnum) },
});

const EventModel = model<IEvent>('Event', EventSchema);

export { EventModel };
