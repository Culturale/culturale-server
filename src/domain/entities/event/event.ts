import type { MongoId } from '~/types/types';

import type { IChat } from '../chat';
import type { IUser } from '../user';

import type { IEvent } from './event.interface';

export type EventProps = {
  id?: MongoId;
  codi: number;
  denominacio: string;
  descripcio: string;
  dataIni: Date;
  dataFi: Date;
  horari: string;
  adress: string;
  url: string;
  chat?: IChat;
  participants?: IUser[];
};

export class Event implements IEvent {
  public id: MongoId;
  public codi: number;
  public denominacio: string;
  public descripcio: string;
  public dataIni: Date;
  public dataFi: Date;
  public horari: string;
  public adress: string;
  public url: string;
  public chat: IChat;
  public participants: IUser[];

  constructor(props: EventProps) {
    const {id, codi, denominacio, descripcio, dataIni, dataFi, horari, adress, url, chat, participants} = props;
    this.id = id;
    this.codi = codi;
    this.denominacio = denominacio;
    this.descripcio = descripcio;
    this.dataIni = dataIni;
    this.dataFi = dataFi;
    this.horari = horari;
    this.adress = adress;
    this.url = url;
    this.chat = chat;
    this.participants = participants || [];
  }
  public updateParticipant(newParticipant: IUser): void {
    const newParticipants = [...this.participants, newParticipant];
    this.participants = newParticipants;
  }
  public get participantsUsernames(): string[] {
    const ids = this.participants.map((participant) => participant.username);
    return ids;
  }
}
