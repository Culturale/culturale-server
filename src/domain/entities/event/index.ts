export * from "./event";
export * from "./event.interface";
export * from "./event.schema";
