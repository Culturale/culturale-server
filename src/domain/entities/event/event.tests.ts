import { Event } from "./event";
import type { IEvent } from "./event.interface";

describe("Event Entity", function () {
  let instance: IEvent;

  beforeEach(function () {
    const date: Date = new Date();
    const number: Number = new Number();
    instance = new Event(
      "test-id",
      "test-denominacio",
      "test-descripcio",
      date,
      date,
      "test-horari",
      "test-adress",
      "test-url",
      number,
      number,
      "test-categoria",
      number,
      number,
      number,
    );
  });

  it("can be created", function () {
    expect(instance).toMatchSnapshot();
  });
});
