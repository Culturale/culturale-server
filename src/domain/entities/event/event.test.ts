import { Types } from 'mongoose';

import type { IChat } from '~/domain/entities/chat/chat.interface';

import { Event } from './event';
import type { IEvent } from './event.interface';

describe('Event Entity', function () {
  let instance: IEvent;

  const dateSpy = jest.spyOn(Date, 'now').mockImplementation(() => 1234);

  beforeEach(function () {
    const time = Date.now();
    const date = new Date(time);
    const chat: IChat = null;
    instance = new Event(
      new Types.ObjectId('642414c21f0577394a8a3099'),
      20211006023,
      'test-denominacio',
      'test-descripcio',
      date,
      date,
      'test-horari',
      'test-adress',
      'test-url',
      chat
    );
  });

  it('can be created', function () {
    expect(instance).toMatchSnapshot();
    dateSpy.mockReset();
  });
});
