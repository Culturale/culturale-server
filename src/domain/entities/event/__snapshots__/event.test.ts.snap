// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Event Entity can be created 1`] = `
Event {
  "adress": "test-adress",
  "chat": undefined,
  "codi": 20211006023,
  "dataFi": 1970-01-01T00:00:01.234Z,
  "dataIni": 1970-01-01T00:00:01.234Z,
  "denominacio": "test-denominacio",
  "descripcio": "test-descripcio",
  "horari": "2h",
  "id": undefined,
  "participants": [],
  "url": "test-adress",
}
`;
